### PR TITLE
Enforce encoded arguments' length

### LINKED ISSUE #403: Enforce encodedArg parameter length as defined in GatewayABIConfig
## Expected behavior
When creating a SideEffect, it should be ensured that the encodedArgs have the length described in the gatewayAbi. An error should be thrown if the length is incorrect.

### Actual behavior
currently, we don't check if the encodedArgs length matches the one specified in the gateway configuration. This can lead to unwanted behavior. For example, codec converts as ss58 address into a 46 bytes (I think) bytes array by default. This then leads to the address being cut of, making it impossible to release the funds at the end of the lifecycle. Essentially, they are now bricked.

This can be ensured by checking that the encoded length matches the one that is stored in `GatewayAbiConfig`. The obvious spot to enforce this would be `types/src/abi.rs:eval_abi()`. If the length check fails, a fitting error should be returned in that function. This automatically reverts the transaction.

### PR DESCRIPTION
**Pull request by palozano**
- refactor: update pallets account-manager src
- refactor: update circuit for safe math
- refactor: update executors for safe math
- refactor: update treasury for safe math
- refactor: update xdns for safe math
- test: update account-manager for safe math test
- chore: missed an update on circuit for safe math test
- chore: missed an update on treasury for safe math test
- fix: update safe math err handling
- chore: update safe math for err handling
- chore: update safe math for err handling in circuit
- chore: update safe math

When creating SideEffect, enforce encoded arguments' length as defined 
in GatewayABIConfig by checking that the (encoded) length matches the 
one in GatewayAbiConfig.

If the length check fails, a fitting error is returned in that function,
reverting the transaction.

Fixes #403

### CODE REVIEW

**Comment by AwesomeIbex** [comment]
this is safemathops no?

File: pallets/circuit/src/optimistic/mod.rs [hunk 1]
```diff
@@ -40,7 +40,12 @@ impl<T: Config> Optimistic<T> {
             }
         }
         // Check if bid candidate has enough balance and reserve
-        <T::Escrowed as EscrowTrait<T>>::Currency::reserve(executor, bid + fsx.input.insurance)
+        let _bid = if let Some(v) = bid.checked_add(&fsx.input.insurance) {
```
Comment (MaciejBaj) [inline pallets/circuit/src/optimistic/mod.rs:43]: no underscore since used

File: pallets/circuit/src/optimistic/mod.rs [hunk 2]
```diff
@@ -116,17 +132,22 @@ impl<T: Config> Optimistic<T> {
                 let (insurance, reserved_bond) =
                     (*sfx_bid.get_insurance(), *sfx_bid.expect_reserved_bond());
 
+                let nsrnc = if let Some(v) = insurance.checked_add(&reserved_bond) {
```
Comment (MaciejBaj) [inline pallets/circuit/src/optimistic/mod.rs:135]: rename to let checked_insurance =

File: pallets/circuit/src/optimistic/mod.rs [hunk 3]
```diff
@@ -147,10 +168,24 @@ impl<T: Config> Optimistic<T> {
                 };
 
                 // First slash executor
-                slashed_reserve += insurance + reserved_bond;
+                if let Some(v) = slashed_reserve
+                    .checked_add(&insurance)
+                    .unwrap()
```
Comment (MaciejBaj) [inline pallets/circuit/src/optimistic/mod.rs:173]: no unwrap

File: pallets/treasury/src/lib.rs [hunk 4]
```diff
@@ -304,23 +305,54 @@ pub mod pallet {
...
             let (count_devs, count_execs) =
-                <Beneficiaries<T>>::iter_keys().fold((0, 0), |mut acc, c| {
+                <Beneficiaries<T>>::iter_keys().fold((0, 0), |mut acc: (i64, i64), c| {
                     match c.1 {
                         BeneficiaryRole::Developer => {
-                            acc.0 += 1;
+                            if let Some(v) = acc.0.checked_add(1) {
+                                acc.0 = v
+                            }
+                            // cannot return Err
```
Comment (MaciejBaj) [inline pallets/treasury/src/lib.rs:314]: create todos and open corresponding issue

File: pallets/treasury/src/lib.rs [hunk 5]
```diff
@@ -304,23 +305,54 @@ pub mod pallet {
...
-            let relative_per_exec =
-                Perbill::from_rational(1, count_execs as u32) * inflation_alloc.executor;
+            let relative_per_dev = if let Some(v) =
+                Perbill::from_rational(1, count_devs as u32).checked_mul(&inflation_alloc.developer)
+            {
+                v
+            } else {
+                Perbill::from_rational(1, count_devs as u32)
+            };
+            // cannot return Err
```
Comment (MaciejBaj) [inline pallets/treasury/src/lib.rs:340]: create todos and open corresponding issue

**Review by MaciejBaj — CHANGES_REQUESTED** [review]

File: pallets/circuit/src/optimistic/mod.rs (hunk 2 above)
Comment (palozano) [reply pallets/circuit/src/optimistic/mod.rs:135]: Done

File: pallets/circuit/src/optimistic/mod.rs (hunk 3 above)
Comment (palozano) [reply pallets/circuit/src/optimistic/mod.rs:173]: replaced it with two if let statements

File: pallets/treasury/src/lib.rs (hunk 4 above)
Comment (palozano) [reply pallets/treasury/src/lib.rs:314]: Added the TODOs and the issue

File: pallets/treasury/src/lib.rs (hunk 5 above)
Comment (palozano) [reply pallets/treasury/src/lib.rs:340]: Added the TODOs and the issue

**Pull request closed without merging**